### PR TITLE
BLIP loss

### PR DESCRIPTION
**Pull request by samedii**
I had to lower num_cuts when running.
```
    prompts="warrior. concept art. trending on artstation",
    drawer="super_resolution",
    size=[512, 512],
    num_cuts=8,
    quality="normal",
    learning_rate=0.1,
    init_image="human.jpg",
```

### CODE REVIEW

**Comment by dribnet** [comment]
Hi - checking this out now but a bit confused on how to get started. Does this branch need the BLIP submodule?

**Comment by samedii** [comment]
Updated with fix for config

**Comment by dribnet** [comment]
been experimenting with this - i've added some downstream commits in the `blip-loss` branch here that allow running blip only with cut_sizes that match blip native inputs. Eg:
```bash
python pixray.py \
   --seed baseline \
   --quality best \
   --size 684 384 \
   --clip_models none \
   --cut_sizes 384 \
   --custom_loss blip \
   --prompts 'sunset at sea' \
   --outdir outputs/feb22/sunset_08_03
```
output



Will followup with experiments but for now wanted you to be aware of this version.

**Comment by dribnet** [comment]
Did testing on various models from the [BLIP README](https://github.com/salesforce/BLIP/blob/main/README.md) table:

### Finetuned checkpoints:
Task | BLIP w/ ViT-B | BLIP w/ ViT-B and CapFilt-L | BLIP w/ ViT-L 
--- | :---: | :---: | :---:
Image-Text Retrieval (COCO) | <a href="https://storage.googleapis.com/sfr-vision-language-research/BLIP/models/model_base_retrieval_coco.pth">Download</a>| - | <a href="https://storage.googleapis.com/sfr-vision-language-research/BLIP/models/model_large_retrieval_coco.pth">Download</a>
Image-Text Retrieval (Flickr30k) | <a href="https://storage.googleapis.com/sfr-vision-language-research/BLIP/models/model_base_retrieval_flickr.pth">Download</a>|  - | <a href="https://storage.googleapis.com/sfr-vision-language-research/BLIP/models/model_large_retrieval_flickr.pth">Download</a>
Image Captioning (COCO) | - | <a href="https://storage.googleapis.com/sfr-vision-language-research/BLIP/models/model*_base_caption.pth">Download</a>| <a href="https://storage.googleapis.com/sfr-vision-language-research/BLIP/models/model_large_caption.pth">Download</a> | 
VQA | <a href="https://storage.googleapis.com/sfr-vision-language-research/BLIP/models/model_vqa.pth">Download</a>| <a href="https://storage.googleapis.com/sfr-vision-language-research/BLIP/models/model*_vqa.pth">Download</a> | - 
NLVR2 | <a href="https://storage.googleapis.com/sfr-vision-language-research/BLIP/models/model_base_nlvr.pth">Download</a>| - | - 

Added new `blip_task` and `blip_size` args to support some others. Here's an example command line using the `coco` / `large` model. 

```bash
python pixray.py \
   --seed baseline \
   --quality normal \
   --num_cuts 8 \
   --batches 3 \
   --size 684 384 \
   --clip_models none \
   --cut_sizes 384 \
   --custom_loss blip \
   --blip_task coco \
   --blip_size large \
   --prompts 'sunset at sea' \
   --outdir outputs/feb22/sunset_09_04
```

And here's some pure blip outputs of "sunset at sea" (a test phrase) just to get a feel for different models. 

The first is `flickr` and `base` (maybe the best of the bunch):
output

`flickr` and `large`:
output

`coco` and `base`
output

`coco` and `large`
output

Now the other models I don't think have an ITM head so were bombing at at model load time. For fun I made the error a warning and surprisingly the code still runs, but I think the optimization is then random. Here's some of those which maybe are just interesting as curiosities or to see vague image quality:

`caption` and `large`
output

`vqa` and `base`
output

`nlvr` and `base`
output

I've again pushed this to the blip-loss branch and made flickr-base the new default.

**Comment by samedii** [comment]
Looks great! :D

**Comment by dribnet** [comment]
I think I'd like to merge my version of this branch into `future` in its current state. Let me know if you have any objections.

**Comment by samedii** [comment]
Go for it :) Nice updates!

**Comment by dribnet** [comment]
Noting that in salesforce/BLIP#4 there is discussion of this PR and @LiJunnan1992 suggests instead perhaps using the raw logit reducing it to minimize the softmax loss.

**Comment by samedii** [comment]
I've tested this a bit before. I got the best results with
```
spherical_distance_itc
```
But I kept the softmax itm loss too in my final loss. I decided to optimize the softmax though since it doesn't behave that well and it felt better to "cap" it's effect.
```
return (spherical_distance_itc + itm_loss) / 2
```
https://github.com/pixray/pixray/blob/5a640eb56446072e41d3b885c3a082cd71588882/Losses/BLIPLoss.py
